pkg/experiment/metastore/metastore_operator.go

```diff
@@ -0,0 +1,90 @@
+package metastore
+
+import (
+	"context"
+	"strconv"
+
+	"connectrpc.com/connect"
+	"github.com/hashicorp/raft"
+
+	metastorev1 "github.com/grafana/pyroscope/api/gen/proto/go/metastore/v1"
+)
+
+func (m *Metastore) Info(_ context.Context, _ *connect.Request[metastorev1.InfoRequest]) (*connect.Response[metastorev1.InfoResponse], error) {
+	cfgFuture := m.raft.GetConfiguration()
+	err := cfgFuture.Error()
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := cfgFuture.Configuration()
+	_, leaderID := m.raft.LeaderWithID()
+	stats := m.raft.Stats()
+
+	res := &metastorev1.InfoResponse{
+		Id:       m.config.Raft.ServerID,
+		State:    metastorev1.State(m.raft.State()),
+		LeaderId: string(leaderID),
+		Term:     getUint64(stats, "term"),
+		Log: &metastorev1.Log{
+			CommitIndex:      m.raft.CommitIndex(),
+			AppliedIndex:     m.raft.AppliedIndex(),
+			LastIndex:        m.raft.LastIndex(),
+			FsmPendingLength: getUint64(stats, "fsm_pending"),
+		},
+		Snapshot: &metastorev1.Snapshot{
+			LastIndex: getUint64(stats, "snapshot_last_index"),
+			LastTerm:  getUint64(stats, "last_snapshot_term"),
+		},
+		Protocol: &metastorev1.Protocol{
+			Version:            getUint64(stats, "protocol_version"),
+			MinVersion:         getUint64(stats, "protocol_max_version"),
+			MaxVersion:         getUint64(stats, "protocol_min_version"),
+			MinSnapshotVersion: getUint64(stats, "snapshot_version_max"),
+			MaxSnapshotVersion: getUint64(stats, "snapshot_version_min"),
+		},
+	}
+
+	switch m.raft.State() {
+	case raft.Leader:
```

Comment:
Probably does not make a big difference here based on the context of how we use it, but this check can return stale information (a non-leader could "think" they are the leader when they are not). 

There is a `VerifyLeader()` method that we can call which will run a sequence of commands through the raft log that will make it a stronger guarantee.

We use it in https://github.com/grafana/pyroscope/blob/26210f593146e4d042f21c6c2c3c2a3482985a45/pkg/experiment/metastore/metastore_readindex.go#L39 where we return an error if we are not the leader but it can be repurposed here

Reply:
In https://github.com/grafana/pyroscope/pull/3544/commits/270171c9e102fa43637efe2008fa77ab8d70c008 I incorporated this check. It's reflected in the `IsLeaderVerified` field of the response. Here's an example output:

```
ID:                  pyroscope-metastore-0.pyroscope-metastore-headless.pyroscope-test.svc.cluster.local.:9099
State:               Follower
State verified:      true
Leader ID:           pyroscope-metastore-1.pyroscope-metastore-headless.pyroscope-test.svc.cluster.local.:9099
Last leader contact: 2024-09-06T17:15:32-07:00
Term:                8
Suffrage:            Voter
Log:
  Commit index:       988
  Applied index:      988
  Last index:         988
  FSM pending length: 0
Snapshot:
  Last index: 0
  Last term:  0
Protocol:
  Version:              3
  Min version:          0
  Max version:          0
  Min snapshot version: 1
  Max snapshot version: 0
Peers:
  ID:       pyroscope-metastore-1.pyroscope-metastore-headless.pyroscope-test.svc.cluster.local.:9099
  Address:  pyroscope-metastore-1.pyroscope-metastore-headless.pyroscope-test.svc.cluster.local.:9099
  Suffrage: Voter

  ID:       pyroscope-metastore-2.pyroscope-metastore-headless.pyroscope-test.svc.cluster.local.:9099
  Address:  pyroscope-metastore-2.pyroscope-metastore-headless.pyroscope-test.svc.cluster.local.:9099
  Suffrage: Voter
```

I have observed this change causes a noticeable increase in latency of the command. I'm not sure if that's an artifact of my local setup or a more systemic one. Probably nothing to worry about now though.

Reply:
The `State verified` field will be true if the state reported by the node matches the state the cluster thinks the node is. I.e. if the node thinks it's a leader and the cluster agrees, then it is set to `true`. If the node thinks it's not a leader, and the cluster agress, then it is set to `true`, etc.

Reply:
Yeah, the latency increase is expected as the call should result in network requests between peers but I am not sure what is the expected order of magnitude for this latency. Probably fine for a command line tool for now.